### DISCUSSION TITLE
Sub volume equivariance

### DISCUSSION THREAD

**Discussion by StefanBst** [e3nn/e3nn#407] [Q&A] [answered]
Dear developers of e3nn,
I was trying to understand and build a 3D equivariant cnn. 
I used 5 GatedBlocks with this configuration: 
GatedBlock(
  (scalar_act): ScalarActivation()
  (gate_act): ScalarActivation()
  (conv): SE3Convolution (SE3Kernel ([(1, 0)] -> [(16, 0), (16, 1), (4, 2), (20, 0)], size=5), kwargs={'stride': 1, 'padding': 2, 'dilation': 1})
)
Now my question would be:
If I put my input volume of size (1,1,40,40,40) into the network and a sub-volume (centre part of the original volume) of size (1,1,20,40,40), the output of the network at the center slice is the same. (out_whole_vol[0,0,20,:,:] == out_sub_vol[0,0,10,:,:])
But if I take a smaller part, say (1,1,5,40,40), the output is different.  (out_whole_vol[0,0,20,:,:] != out_sub_vol[0,0,3,:,:])
Shouldn't the output at the centre slice be the same if the input size >= kernel size? (in this case 5)

Thank you :)

**Comment by mariogeiger** [comment]
Can you try with 2 instead of 3?

**Reply by StefanBst** [reply]
Thank you for your quick answer!
the result is the same with 2 instead of 3.
I did some testing with (1,1,40,40,40) and  (1,1,6,40,40)  and (out_whole_vol[0,0,20,:,:] != out_sub_vol[0,0,3,:,:])
For 1 GatedBlock -3/+3: 
(1,1,40,40,40) and (1,1,6,40,40) : out_whole_vol[0,0,20,:,:] == out_sub_vol[0,0,3,:,:]

5 GatedBlocks in Sequence dif -3:+3:
(1,1,40,40,40) and (1,1,6,40,40) : out_whole_vol[0,0,20,:,:] != out_sub_vol[0,0,3,:,:]

5 GatedBlocks in sequence dif -10:+10
(1,1,40,40,40) and (1,1,20,40,40) : out_whole_vol[0,0,20,:,:] != out_sub_vol[0,0,10,:,:]

5 GatedBlocks in sequence dif -15:+15
(1,1,40,40,40) and (1,1,30,40,40) : out_whole_vol[0,0,20,:,:] == out_sub_vol[0,0,15,:,:]

Here is a link to Live Notion (https://www.notion.so/) with a more detailed description and visualisation of each slice and their similarity:
https://aspiring-grease-7de.notion.site/Sub_volume_test-d68443fc621b488ba1a4259bb8f3102c

Thank you!

maybe important:
First test: 
GatedBlock(
  (scalar_act): ScalarActivation()
  (gate_act): ScalarActivation()
  (conv): SE3Convolution (SE3Kernel ([(1, 0)] -> [(16, 0), (16, 1), (4, 2), (20, 0)], size=5), kwargs={'stride': 1, 'padding': 2, 'dilation': 1})
)
Other tests:
[GatedBlock(
   (scalar_act): ScalarActivation()
   (gate_act): ScalarActivation()
   (conv): SE3Convolution (SE3Kernel ([(1, 0)] -> [(16, 0), (16, 1), (4, 2), (20, 0)], size=5), kwargs={'stride': 1, 'padding': 2, 'dilation': 1})
 ),
 GatedBlock(
   (scalar_act): ScalarActivation()
   (gate_act): ScalarActivation()
   (conv): SE3Convolution (SE3Kernel ([(16, 0), (16, 1), (4, 2)] -> [(16, 0), (16, 1), (4, 2), (20, 0)], size=5), kwargs={'stride': 1, 'padding': 2, 'dilation': 1})
 ),
 GatedBlock(
   (scalar_act): ScalarActivation()
   (gate_act): ScalarActivation()
   (conv): SE3Convolution (SE3Kernel ([(16, 0), (16, 1), (4, 2)] -> [(16, 0), (16, 1), (4, 2), (20, 0)], size=5), kwargs={'stride': 1, 'padding': 2, 'dilation': 1})
 ),
 GatedBlock(
   (scalar_act): ScalarActivation()
   (gate_act): ScalarActivation()
   (conv): SE3Convolution (SE3Kernel ([(16, 0), (16, 1), (4, 2)] -> [(16, 0), (16, 1), (4, 2), (20, 0)], size=5), kwargs={'stride': 1, 'padding': 2, 'dilation': 1})
 ),
 GatedBlock(
   (scalar_act): ScalarActivation()
   (conv): SE3Convolution (SE3Kernel ([(16, 0), (16, 1), (4, 2)] -> [(64, 0)], size=5), kwargs={'stride': 1, 'padding': 2, 'dilation': 1})
 )]

**Reply by mariogeiger** [reply, marked as answer]
I guess it's simply the border effects. With 5 convolutions of kernel_size=5 you have a perception field of 10 pixels away (5 * 5//2).
Therefore anything that is smaller to 21 pixels will show this discrepancy.

Btw you should be more careful when you say -3:+3 and use an input of size 6 and look at position 3. Technically you are doing -3:+2 here

**Reply by StefanBst** [reply]
Thank you so much for the fast and clear answers!
